utils/staking-miner/src/monitor.rs

```diff
@@ -0,0 +1,167 @@
+	}
+}
+
+/// Ensure that our current `us` have not submitted anything previously.
+async fn ensure_no_previous_solution<
+	T: EPM::Config + frame_system::Config<AccountId = AccountId>,
+	B: BlockT,
+>(
+	ext: &mut Ext,
+	us: &AccountId,
+) -> Result<(), Error> {
+	use EPM::signed::SignedSubmissions;
+	ext.execute_with(|| {
+		if <SignedSubmissions<T>>::get().iter().any(|ss| &ss.who == us) {
+			Err(Error::AlreadySubmitted)
+		} else {
+			Ok(())
+		}
+	})
+}
+
+macro_rules! monitor_cmd_for { ($runtime:tt) => { paste::paste! {
+	/// The monitor command.
+	pub(crate) async fn [<monitor_cmd_ $runtime>](
+		client: WsClient,
+		shared: SharedConfig,
+		config: MonitorConfig,
+		signer: Signer,
+	) -> Result<(), Error> {
+		use $crate::[<$runtime _runtime_exports>]::*;
+		let (sub, unsub) = if config.listen == "head" {
+			("chain_subscribeNewHeads", "chain_unsubscribeNewHeads")
+		} else {
+			("chain_subscribeFinalizedHeads", "chain_unsubscribeFinalizedHeads")
+		};
+
+		log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", sub, unsub);
+		let mut subscription: Subscription<Header> = client
+			.subscribe(&sub, JsonRpcParams::NoParams, &unsub)
+			.await
+			.unwrap();
+
+		while let Some(now) = subscription.next().await.unwrap() {
+			let hash = now.hash();
+			log::debug!(target: LOG_TARGET, "new event at #{:?} ({:?})", now.number, hash);
+
+			// we prefer doing this check before fetching anything into a remote-ext.
+			if ensure_signed_phase::<Runtime, Block>(&client, hash).await.is_err() {
+				log::debug!(target: LOG_TARGET, "phase closed, not interested in this block at all.");
+				continue;
+			};
+
+			// NOTE: we don't check the score of any of the submitted solutions. If we submit a weak
+			// one, as long as we are valid, we will end up getting our deposit back, so not a big
+			// deal for now. Note that to avoid an unfeasible solution, we should make sure that we
+			// only start the process on a finalized snapshot. If the signed phase is long enough,
+			// this will not be a solution.
+
+			// grab an externalities without staking, just the election snapshot.
+			let mut ext = crate::create_election_ext::<Runtime, Block>(shared.uri.clone(), Some(hash), false).await?;
+
+			if ensure_no_previous_solution::<Runtime, Block>(&mut ext, &signer.account).await.is_err() {
+				log::debug!(target: LOG_TARGET, "We already have a solution in this phase, skipping.");
+				continue;
+			}
+
+			let (raw_solution, witness) = crate::mine_checked::<Runtime>(&mut ext, 50)?;
+			log::info!(target: LOG_TARGET, "mined solution with {:?}", &raw_solution.score);
+
+			let nonce = crate::get_account_info::<Runtime>(&client, &signer.account, Some(hash))
+				.await?
+				.map(|i| i.nonce)
+				.expect("signer account is checked to exist upon startup; it can only die if it \
+					transfers funds out of it, or get slashed. If it does not exist at this point, \
+					it is likely due to a bug, or the signer got slashed. Terminating."
+				);
```

Comment:
This code, including the long expect, exists elsewhere in the codebase. Probably better to extract a function instead of copy-pasting it.